rotkehlchen/db/dbhandler.py

```diff
@@ -2233,6 +2234,95 @@ def get_netvalue_data(
                 data.append(total)
         return times_int, data
 
+    @staticmethod
+    def _infer_zero_timed_balances(
+            cursor: 'DBCursor',
+            balances: list[SingleDBAssetBalance],
+            from_ts: Optional[Timestamp] = None,
+            to_ts: Optional[Timestamp] = None,
+    ) -> list[SingleDBAssetBalance]:
+        """
+        Given a list of asset specific timed balances, infers the missing zero timed balances
+        for the asset. We add 0 balances on the start and end of a period of 0 balances.
+
+        Example
+        We have the following timed balances for ETH (value, time):
+        (1, 1), (1, 2), (2, 3), (5, 7), (5, 12)
+        The timestamps of all timed balances in the DB are:
+        (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
+        So we need to infer the following zero timed balances:
+        (0, 4), (0, 6), (0, 8), (0, 11)
+        """
+        if len(balances) == 0:
+            return []
+
+        if from_ts is None:
+            from_ts = Timestamp(0)
+        if to_ts is None:
+            to_ts = ts_now()
+
+        # else
+        cursor.execute(
+            'SELECT COUNT(*) FROM timed_balances WHERE timestamp BETWEEN ? AND ?',
+            (from_ts, to_ts),
+        )
+        num_all_balances = cursor.fetchone()[0]
+
+        if len(balances) == num_all_balances:
+            return []
+
+        # else
+        asset_timestamps = [b.time for b in balances]
+        cursor.execute(
+            'SELECT timestamp, category FROM timed_balances WHERE timestamp BETWEEN ? AND ? '
+            'ORDER BY timestamp ASC',
+            (from_ts, to_ts),
+        )
+        all_timestamps, all_categories = zip(*cursor)
+        # dicts maintain insertion order in python 3.7+
+        timestamps_have_asset_balance = dict.fromkeys(all_timestamps, False)
+        timestamps_with_asset_balance = dict.fromkeys(asset_timestamps, True)
+        timestamps_have_asset_balance.update(timestamps_with_asset_balance)
+        prev_has_asset_balance = timestamps_have_asset_balance[all_timestamps[0]]  # the value of the first timestamp  # noqa: E501
+        prev_timestamp = all_timestamps[0]
+        inferred_balances: list[SingleDBAssetBalance] = []
+        for idx, (timestamp, has_asset_balance) in enumerate(timestamps_have_asset_balance.items()):  # noqa: E501
+            if idx == len(timestamps_have_asset_balance) - 1 and has_asset_balance is False:
+                # If there is no balance for the last timestamp add a zero balance.
+                inferred_balances.append(
+                    SingleDBAssetBalance(
+                        time=timestamp,
+                        amount=ZERO,
+                        usd_value=ZERO,
+                        # the category of the asset at the start of the zero balance period
+                        category=inferred_balances[-1].category,
```

Comment:
Can't this raise error if you have two snapshots and in the second one you are missing the asset? Wouldn't `inferred_balances` be empty?